chain/catchup.go

```diff
@@ -0,0 +1,123 @@
+package chain
+
+import (
+	"context"
+	"time"
+
+	// :(
+	logging "github.com/ipfs/go-log"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var logCatchup = logging.Logger("chain/catchup")
+
+type retryableError struct {
+	Err error
+}
+
+func (re retryableError) Error() string { return re.Err.Error() }
+
+func isRetryError(err error) bool {
+	if err == nil {
+		return false
+	}
+	_, ok := err.(retryableError)
+	return ok
+}
+
+type catchupStore interface {
+	GetHead() types.TipSetKey
+	GetTipSet(tsKey types.TipSetKey) (types.TipSet, error)
+}
+
+type catchupTracker interface {
+	SelectHead() (*types.ChainInfo, error)
+	UpdateTrusted(context.Context) error
+}
+
+type catchupSyncer interface {
+	HandleNewTipSet(context.Context, *types.ChainInfo, bool) error
+	Status() Status
+}
+
+type catchupComplete func(newHead *types.ChainInfo, curHead types.TipSet) (bool, error)
+
+type CatcherUpper struct {
+	tracker catchupTracker
+	syncer  catchupSyncer
+	store   catchupStore
+	done    catchupComplete
+}
+
+// NewCatcherUpper is a poorly named method that returns a CatcherUpper.
+func NewCatcherUpper(t catchupTracker, sy catchupSyncer, st catchupStore, done catchupComplete) *CatcherUpper {
+	return &CatcherUpper{
+		tracker: t,
+		syncer:  sy,
+		store:   st,
+		done:    done,
+	}
+}
+
+// Start starts a go routine to performe the catchup process.
+func (cu *CatcherUpper) Start(ctx context.Context, tick clockwork.Ticker) <-chan error {
+	logCatchup.Info("starting cahin catchup runner")
+	errors := make(chan error)
+	go func() {
+		for {
+			select {
+			case <-tick.Chan():
+				complete, err := cu.tryCatchUp(ctx)
+				if err != nil && !isRetryError(err) {
+					errors <- err
+					logCatchup.Errorf("chain catchup encountered unretryable error: %s", err.Error())
+					return
+				}
+				if isRetryError(err) {
+					logCatchup.Warnf("chain catchup encountered retryable error: %s", err.Error())
+					continue
+				}
+				if complete {
+					logCatchup.Info("completed chain catchup status: %s", cu.syncer.Status().String())
+					return
+				}
+				logCatchup.Infof("chain catchup status: %s", cu.syncer.Status().String())
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return errors
+}
+
+func (cu *CatcherUpper) tryCatchUp(ctx context.Context) (bool, error) {
+	// It bothers me that tick could have a shorter duration than the timeout...
+	updtCtx, cancel := context.WithTimeout(ctx, time.Second*15)
+	defer cancel()
+
+	// if this failes it means we are tracking peers, but none of them replied.
+	if err := cu.tracker.UpdateTrusted(updtCtx); err != nil {
+		return false, retryableError{err}
```

Comment:
Borrowed this error wrapping from [os.errors](https://github.com/golang/go/blob/master/src/os/error.go#L110). Any errors that come from the tracker should be retried. Errors from the store and syncer indicate a more serious problem.

Reply:
I think it would be simpler to just return three values: success, retry, error